sdk/digitaltwins/digitaltwins/src/digitalTwinsClient.ts

```diff
@@ -0,0 +1,898 @@
+// Licensed under the MIT License.
+
+/// <reference lib="esnext.asynciterable" />
+
+import {
+  TokenCredential,
+  RestResponse,
+  RequestOptionsBase,
+  OperationOptions,
+  bearerTokenAuthenticationPolicy,
+  createPipelineFromOptions,
+  generateUuid
+} from "@azure/core-http";
+import { PageSettings, PagedAsyncIterableIterator } from "@azure/core-paging";
+import { AzureDigitalTwinsAPI as GeneratedClient } from "./generated/azureDigitalTwinsAPI";
+import {
+  AzureDigitalTwinsAPIOptionalParams,
+  DigitalTwinsGetByIdResponse,
+  DigitalTwinsAddResponse,
+  DigitalTwinsUpdateOptionalParams,
+  DigitalTwinsUpdateResponse,
+  DigitalTwinsDeleteOptionalParams,
+  DigitalTwinsGetComponentResponse,
+  DigitalTwinsUpdateComponentResponse,
+  DigitalTwinsUpdateComponentOptionalParams,
+  DigitalTwinsAddRelationshipResponse,
+  DigitalTwinsAddRelationshipOptionalParams,
+  DigitalTwinsUpdateRelationshipOptionalParams,
+  DigitalTwinsUpdateRelationshipResponse,
+  DigitalTwinsDeleteRelationshipOptionalParams,
+  DigitalTwinsSendTelemetryOptionalParams,
+  DigitalTwinsSendComponentTelemetryOptionalParams,
+  DigitalTwinsListRelationshipsResponse,
+  IncomingRelationship,
+  DigitalTwinsListIncomingRelationshipsResponse,
+  ModelData,
+  DigitalTwinModelsGetByIdResponse,
+  DigitalTwinModelsGetByIdOptionalParams,
+  DigitalTwinModelsAddResponse,
+  DigitalTwinModelsAddOptionalParams,
+  DigitalTwinModelsListResponse,
+  DigitalTwinModelsListOptionalParams,
+  DigitalTwinModelsListOptions,
+  EventRoutesGetByIdResponse,
+  EventRoute,
+  EventRoutesAddOptionalParams,
+  EventRoutesListNextResponse,
+  EventRoutesListOptionalParams,
+  QueryQueryTwinsResponse,
+  QuerySpecification
+} from "./generated/models";
+
+/**
+ * Client for Azure IoT DigitalTwins API.
+ */
+export class DigitalTwinsClient {
+  /**
+   * @internal
+   * @ignore
+   * A reference to the auto-generated AzureDigitalTwinsAPI
+   */
+  private readonly client: GeneratedClient;
+
+  // private readonly client!: AzureDigitalTwinsAPI;
+
+  /**
+   * Creates an instance of AzureDigitalTwinsAPI.
+   *
+   * Example usage:
+   * ```ts
+   * const { DigitalTwinsClient, ServiceClientCredentials } = require("@azure/digitaltwins");
+   *
+   * const client = new DigitalTwinsClient(
+   *   "<endpoint>",
+   *   new DefaultAzureCredential();
+   * );
+   * ```
+   * @param {string} endpoint The endpoint of the service. If options parameter is given than this value will be ignored.
+   * @param {TokenCredential} credential Used to authenticate requests to the service.
+   * @param {AzureDigitalTwinsAPI} [options] Used to configure the service client.
+   */
+  constructor(
+    endpoint: string,
+    credential: TokenCredential,
+    options: AzureDigitalTwinsAPIOptionalParams = {}
+  ) {
+    const authPolicy = bearerTokenAuthenticationPolicy(
+      credential,
+      "https://digitaltwins.azure.net/.default"
+    );
+    this.client = new GeneratedClient(credential, {
```

Comment:
you can set `add-credentials: false` in the swagger codegen and avoid needing to pass the credential here - you already have a custom pipeline so it's not needed

Reply:
The current implementation is a workaround of the bug:
[https://github.com/Azure/azure-sdk-for-js/issues/9583](https://github.com/Azure/azure-sdk-for-js/issues/9583)

Let's discuss the alternatives offline.


---
In reply to: [470201328](https://github.com/Azure/azure-sdk-for-js/pull/10560#discussion_r470201328) [](ancestors = 470201328)

Reply:
This isn't relevant, that's for when ServiceClient has to add the auth policy, but you're using `createPipelineFromOptions` which makes us take this branch in the code:

https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-http/src/serviceClient.ts#L199

So you don't need to worry about `ServiceClient` making the wrong `bearerTokenAuthenticationPolicy` because *you* are providing it above.

Reply:
Fixed.

---
In reply to: [472563168](https://github.com/Azure/azure-sdk-for-js/pull/10560#discussion_r472563168) [](ancestors = 472563168,470201328)

Reply:
Not it has pushed.

---
In reply to: [474964617](https://github.com/Azure/azure-sdk-for-js/pull/10560#discussion_r474964617) [](ancestors = 474964617)